src/middleware/warnMutations.js

```diff
@@ -0,0 +1,54 @@
+import isEqual from 'lodash/lang/isEqual';
+import any from 'lodash/collection/any';
+import cloneDeep from 'lodash/lang/cloneDeep';
+
+function copyState(state) {
+  return cloneDeep(state);
+}
+
+function wasMutated(prevStateRef, prevState, stateRef, state) {
+  if (prevStateRef === stateRef && !isEqual(prevState, state)) {
+    return true;
+  }
+
+  if (typeof prevStateRef !== 'object') {
+    return false;
+  }
+
+  return any(prevStateRef, (val, key) =>
+    wasMutated(val, prevState[key], stateRef[key], state[key]));
+}
+
+export default function warnMutationsMiddleware(getState) {
+  let lastStateRef = getState();
+  let lastState = copyState(lastStateRef);
+
+  return (next) => (action) => {
+    const stateRef = getState();
+    const state = copyState(stateRef);
+
+    if (wasMutated(lastStateRef, lastState, stateRef, state)) {
+      console.warn([
+        'A state mutation was detected between dispatches.' +
+        ' This may cause incorrect behavior.',
+        '(https://github.com/gaearon/redux#my-views-arent-updating)'
+      ].join(''));
+    }
+
+    action = next(action);
+
+    lastStateRef = getState();
+    lastState = copyState(lastStateRef);
+
+    if (wasMutated(stateRef, state, lastStateRef, lastState)) {
+      console.warn(
```

Comment:
I would prefer to use invariant. If I am using this middleware, I would prefer to get a hard error. This middleware isn't used by default, so it won't cause issues in prod.